ocaml/xapi/xapi_host_helpers.ml

```diff
@@ -308,50 +308,67 @@ module Configuration = struct
             Event_types.event_from_of_rpc
           in
-          List.iter (fun ev ->
-              match Event_helper.record_of_event ev with
-                | Event_helper.Host (host_ref, Some host_rec) -> begin
-                    let oc = host_rec.API.host_other_config in
-                    let iscsi_iqn = try Some (List.assoc "iscsi_iqn" oc) with _ -> None in
-                    begin match iscsi_iqn with
-                      | None -> ()
-                      | Some "" -> ()
-                      | Some iqn when iqn <> host_rec.API.host_iscsi_iqn ->
-                        Client.Client.Host.set_iscsi_iqn rpc session_id host_ref iqn
-                      | _ -> ()
-                    end;
-                    (* Accepted values are "true" and "false" *)
-                    (* If someone deletes the multipathing other_config key, we don't do anything *)
-                    let multipathing = try Some (List.assoc "multipathing" oc |> Pervasives.bool_of_string) with _ -> None in
-                    begin match multipathing with
-                      | None -> ()
-                      | Some multipathing when multipathing <> host_rec.API.host_multipathing ->
-                        Client.Client.Host.set_multipathing rpc session_id host_ref multipathing
-                      | _ -> ()
-                    end
-                  end
-                | _ -> ())
-              events.Event_types.events;
-          events.Event_types.token)
+          let check_host (host_ref,host_rec) =
+            let oc = host_rec.API.host_other_config in
+            let iscsi_iqn = try Some (List.assoc "iscsi_iqn" oc) with _ -> None in
+            begin match iscsi_iqn with
+              | None -> ()
+              | Some "" -> ()
+              | Some iqn when iqn <> host_rec.API.host_iscsi_iqn ->
+                Client.Client.Host.set_iscsi_iqn rpc session_id host_ref iqn
+              | _ -> ()
+            end;
+            (* Accepted values are "true" and "false" *)
+            (* If someone deletes the multipathing other_config key, we don't do anything *)
+            let multipathing = try Some (List.assoc "multipathing" oc |> Pervasives.bool_of_string) with _ -> None in
+            begin match multipathing with
+              | None -> ()
+              | Some multipathing when multipathing <> host_rec.API.host_multipathing ->
+                Client.Client.Host.set_multipathing rpc session_id host_ref multipathing
+              | _ -> ()
+            end
+          in
+          let event_recs = List.map Event_helper.record_of_event events.Event_types.events in
+          let in_rpu = List.fold_left (fun in_rpu ev ->
+              match ev with
+              | Event_helper.Pool (_pool_ref, Some pool_rec) ->
+                let in_rpu = Helpers.rolling_upgrade_in_progress_of_oc pool_rec.API.pool_other_config in
+                if (not in_rpu) && was_in_rpu then begin
```

Comment:
I am a bit confused on how `was_in_rpu` is set. RPU involves restarting XAPI, so if you got a 1 host pool will RPU be reported as in progress on startup, or will it immediately finish because it was the only host to be upgraded?

Reply:
https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/db_gc.ml#L119-L168

Reply:
This code doesn't care if it misses a transient RPU mode since the code it's invoking is entirely level-based rather than edge-based. All we're trying to do is make sure that if we do any suppression at all then we ensure at the end we don't miss anything by updating all hosts.